src/box/memtx_space.c

```diff
@@ -1321,7 +1321,16 @@ memtx_space_build_index(struct space *src_space, struct index *new_index,
 
 		trigger_create(&on_replace, memtx_build_on_replace, &state,
 			       NULL);
-		trigger_add(&src_space->on_replace, &on_replace);
+		/*
+		 * If MVCC is enabled, all concurrent writes that we haven't
+		 * read will be conflicted, so we don't need to set the trigger.
+		 * Moreover, concurrent transaction can be rolled back much
+		 * later than the index will be built so the memtx_ddl_state
+		 * will be invalid in this case since it's allocated on stack
+		 * of this function.
+		 */
+		if (!memtx_tx_manager_use_mvcc_engine)
+			trigger_add(&src_space->on_replace, &on_replace);
```

Comment:
It always conflicted, even before this change. Moreover, it was likely to crash.
Here is the issue to make it work without conflicts: https://github.com/tarantool/tarantool/issues/10170.

Reply:
Also, I suggested to disable background build when MVCC is enabled so that DDL won't be conflicted by a random DML operation on the same, but Sergey Petrenko and I decided that it's actually a bad idea.